{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Modify Nodes in an NDEx Network Based on a TSV File"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook, you will modify a network from NDEx that shows the interactions between the NCI ITCR projects.\n",
    "\n",
    "This version is missing some node attributes and needs corrections to the organization of their attributes.\n",
    "\n",
    "The network will be updated based on a TSV file where each row has information about one of the ITCR projects.\n",
    "\n",
    "The workflow:\n",
    " * Load the TSV file as a Pandas Dataframe\n",
    " * Create an NiceCX network from the network on NDEx\n",
    " * For each row in the Dataframe:\n",
    "  * Identify the corresponding node (skip if there is no such node)\n",
    "  * Add attributes to the node based on the following columns:\n",
    "   * “Category”\n",
    "   * “Tool URL”\n",
    "   * “Code Repo”\n",
    "   * “Project Info” \n",
    "  * Set the \"Tool Description\" attribute of the node to the value of the \"Body\" column\n",
    "  * Set the node name to be the value of the \"Title\" column\n",
    "  * Set the \"label\" attribute to the value of the \"Tool Acronym\" when available, otherwise use the value of the \"Title\" column.\n",
    " * Save the network to your account on NDEx\n",
    "\n",
    "The final network should still have 36 nodes/81 edges but provide extra info and link out to external resource"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, import the ndex2 Python client, Pandas, and itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ndex2\n",
    "import pandas as pd\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, set a variable to hold the UUID of the network we want to modify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "my_network_uuid = '04c0a7e8-af92-11e7-94d3-0ac135e8bacf'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set variables to hold your account and password "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "my_server = 'public.ndexbio.org'\n",
    "my_account = \"drh\"\n",
    "my_password = \"drh\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a NiceCX network from the ITCR network.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloading network and buiding NiceCX...\n",
      "done\n",
      "Name: ITCR Connectivity Map\n",
      "Nodes: 36\n",
      "Edges: 81\n",
      "Node Attributes: 0\n",
      "Edge Attributes: 241\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"downloading network and buiding NiceCX...\")\n",
    "itcr_network = ndex2.create_nice_cx_from_server(server=my_server, uuid=my_network_uuid)\n",
    "#my_network = ndex2.create_nice_cx_from_server(server='public.ndexbio.org', uuid=my_network_uuid, account=my_account, password = my_password)\n",
    "print(\"done\")\n",
    "print(itcr_network.get_summary())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the TSV to a Datframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Title</th>\n",
       "      <th>Tool Acronym</th>\n",
       "      <th>Body</th>\n",
       "      <th>Category</th>\n",
       "      <th>Tool URL</th>\n",
       "      <th>Code Repo</th>\n",
       "      <th>Project Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3D Slicer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3D Slicer is the free open source software for...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>https://www.slicer.org</td>\n",
       "      <td>https://github.com/Slicer</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Allele-Specific Alternative mRNA processing (A...</td>\n",
       "      <td>ASARP</td>\n",
       "      <td>A software pipeline for prediction of allele-s...</td>\n",
       "      <td>omics</td>\n",
       "      <td>https://www.ibp.ucla.edu/research/xiao/Softwar...</td>\n",
       "      <td>https://github.com/cyruschan/ASARP</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Apache Clinical Text and Knowledge Extraction ...</td>\n",
       "      <td>cTAKES</td>\n",
       "      <td>The tool extracts deep phenotypic information ...</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>http://ctakes.apache.org</td>\n",
       "      <td>http://ctakes.apache.org/downloads.cgi</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bioconductor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bioconductor provides tools for the analysis a...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>https://www.bioconductor.org</td>\n",
       "      <td>http://bioconductor.org/developers/how-to/git/</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cancer and Phenomics Toolkit (CaPTK)</td>\n",
       "      <td>CaPTK</td>\n",
       "      <td>CaPTk is a tool that facilitates translation o...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://captk.projects.nitrc.org</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cancer Slide Digital Archive (CDSA)</td>\n",
       "      <td>CDSA</td>\n",
       "      <td>The CDSA is a web-based platform to support th...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://cancer.digitalslidearchive.net</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cancer-Related Analysis of Variants Toolkit (C...</td>\n",
       "      <td>CRAVAT</td>\n",
       "      <td>CRAVAT is an easy to use web-based tool for an...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://www.cravat.us/CRAVAT/</td>\n",
       "      <td>https://github.com/KarchinLab/CRAVAT</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Cistrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Curated and processed human/mouse ChIP/DNase-s...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://cistrome.org/Cistrome/Cistrome_Project....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Clinical Interpretation of Variants in Cancer ...</td>\n",
       "      <td>CIViC</td>\n",
       "      <td>CIViC is an open access, open source, communit...</td>\n",
       "      <td>Omics, Clinical</td>\n",
       "      <td>https://civic.genome.wustl.edu/home</td>\n",
       "      <td>https://github.com/griffithlab</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Cores Of Recurrent Events (CORE)</td>\n",
       "      <td>CORE</td>\n",
       "      <td>CORE is a statistically supported computationa...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://cran.us.r-project.org/web/packages/CORE/</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>D2Refine</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D2Refine is an integrated platform that enable...</td>\n",
       "      <td>Data Standards</td>\n",
       "      <td>https://github.com/caCDE-QA/D2Refine/wiki</td>\n",
       "      <td>https://github.com/caCDE-QA/D2Refine</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DICOM ToolKit (DCMTK)</td>\n",
       "      <td>DCMTK</td>\n",
       "      <td>DCMTK is an open source DICOM toolkit.\\r\\n</td>\n",
       "      <td>Imaging, Data Standards</td>\n",
       "      <td>http://dicom.offis.de/dcmtk.php.en</td>\n",
       "      <td>http://git.dcmtk.org</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Enhancer Linking by Methylation/Expression Rel...</td>\n",
       "      <td>ELMER</td>\n",
       "      <td>R tool for analysis of DNA methylation and exp...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://bioconductor.org/packages/release/bioc/...</td>\n",
       "      <td>https://github.com/lijingya/ELMER</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Feature DB</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The backend database to manage segmentation re...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://github.com/SBU-BMI/pathomics_featuredb</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Galaxy-P multiOmics</td>\n",
       "      <td>Galaxy-P</td>\n",
       "      <td>A unified platform for integrative genomic-pro...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://galaxyp.org</td>\n",
       "      <td>https://github.com/galaxyproteomics/tools-galaxyp</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GenePattern</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GenePattern is a genomic analysis platform tha...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://software.broadinstitute.org/cancer/soft...</td>\n",
       "      <td>https://github.com/genepattern/genepattern-server</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Genome-Independent Identification of RNA Editi...</td>\n",
       "      <td>GIREMI</td>\n",
       "      <td>Identification of single-nucleotide variants i...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://github.com/zhqingit/giremi</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Integrative Genomics Viewer (IGV)</td>\n",
       "      <td>IGV</td>\n",
       "      <td>The Integrative Genomics Viewer (IGV) is a hig...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://www.igv.org</td>\n",
       "      <td>https://github.com/igvteam</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>LSTM-CRF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LSTM-CRF uses Natural Language Processing meth...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://github.com/abhyudaynj/LSTM-CRF-models</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Methylation Integration (Mint)</td>\n",
       "      <td>Mint</td>\n",
       "      <td>A comprehensive software pipeline and set of G...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>https://github.com/sartorlab/mint</td>\n",
       "      <td>https://github.com/sartorlab/mint</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Modified UMS, Modified SemRep and SemMedDB-UTH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Modified UMLS, modified SemRep and SemMedDB-UT...</td>\n",
       "      <td>Clinical, Omics</td>\n",
       "      <td>https://skr3.nlm.nih.gov/SemMedDB/index_uth.html</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Mutated Positions in Annotated Structures (MuPIT)</td>\n",
       "      <td>MuPIT</td>\n",
       "      <td>MuPIT provides 3D visualization of mutation cl...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://mupit.icm.jhu.edu/MuPIT_Interactive/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>The Network Data Exchange (NDEx)</td>\n",
       "      <td>NDEx</td>\n",
       "      <td>NDEx is an online commons where scientists can...</td>\n",
       "      <td>Network Biology</td>\n",
       "      <td>http://www.ndexbio.org/</td>\n",
       "      <td>https://github.com/ndexbio</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Next Generation Clustered Heat Maps</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Next-Generation (Clustered) Heat Maps are inte...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://bioinformatics.mdanderson.org/TCGA/NGCH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>OmniSearch</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OmniSearch is a semantic search software based...</td>\n",
       "      <td>Data Standards</td>\n",
       "      <td>http://omnisearch.soc.southalabama.edu/ui/</td>\n",
       "      <td>https://github.com/OmniSearch/ncro</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Ontology for MIcroRNA Target (OMIT)</td>\n",
       "      <td>OMIT</td>\n",
       "      <td>OMIT is a domain ontology specifically designe...</td>\n",
       "      <td>Data Standards</td>\n",
       "      <td>http://omnisearch.soc.southalabama.edu/w/index...</td>\n",
       "      <td>https://github.com/OmniSearch/ncro</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>LesionTracker Web Image Viewer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Extensible open-source zero-footprint web imag...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://lesiontracker.ohif.org</td>\n",
       "      <td>https://github.com/OHIF</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>owl-qa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>owl-qa is an OWL-based QA tool for cancer stud...</td>\n",
       "      <td>Data Standards</td>\n",
       "      <td>https://github.com/caCDE-QA/owl-qa</td>\n",
       "      <td>https://github.com/caCDE-QA/owl-qa</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P-Mart</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Web-based software platform to enable analysis...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>https://pmart.labworks.org/index.html</td>\n",
       "      <td>https://hub.docker.com/r/pnnl/pmart-web/</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Personalized Cancer Therapy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Curated website cataloguing clinically actiona...</td>\n",
       "      <td>Clinical, Omics</td>\n",
       "      <td>https://pct.mdanderson.org/#/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>PhenStat</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PhenStat is a freely available R package that ...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://bioconductor.org/packages/release/bioc...</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Quantitative Imaging In Pathology (QUIP)</td>\n",
       "      <td>QUIP</td>\n",
       "      <td>This site hosts web accessible applications, t...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://quip.bmi.stonybrook.edu</td>\n",
       "      <td>https://github.com/SBU-BMI/nscale</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>RADIOMICS.io</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RADIOMICS.io is a open source platform for inf...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>https://www.radiomics.io</td>\n",
       "      <td>https://github.com/Radiomics</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Region Templates</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A suite of libraries and runtime to support im...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://github.com/SBU-BMI/region-templates</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SampleMap</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Projects samples into a GoogleMap explorer to ...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>https://tumormap.ucsc.edu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Pathology Imaging Informatics Platform (PIIP)</td>\n",
       "      <td>PIIP</td>\n",
       "      <td>Freely downloadable pathology viewer which sup...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://pathiip.org</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>FHIR ShEx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ShEx is a QA tool based on the Shape Expressio...</td>\n",
       "      <td>Data Standards</td>\n",
       "      <td>https://github.com/caCDE-QA/ShEX</td>\n",
       "      <td>https://github.com/caCDE-QA/ShEX</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>SlicerDMRI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Diffusion magnetic resonance imaging for neuro...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://dmri.slicer.org</td>\n",
       "      <td>https://github.com/SlicerDMRI</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Text Information Extraction System (TIES)</td>\n",
       "      <td>TIES</td>\n",
       "      <td>TIES is a service based software system for ac...</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>http://ties.dbmi.pitt.edu</td>\n",
       "      <td>https://sourceforge.net/projects/caties/</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>The Laboratory for Individualized Breast Radio...</td>\n",
       "      <td>LIBRA</td>\n",
       "      <td>LIBRA is a fully-automatic breast density esti...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>https://www.med.upenn.edu/sbia/libra.html</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Trinity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Trinity Cancer Transcriptome Analysis Toolkit ...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>https://github.com/trinityrnaseq/trinityrnaseq...</td>\n",
       "      <td>https://github.com/trinityrnaseq/trinityrnaseq</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>UCSC Xena</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UCSC Xena securely analyzes and visualizes you...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://xena.ucsc.edu</td>\n",
       "      <td>https://github.com/ucscXena/</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>WebMeV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Web MeV (Multiple-experiment Viewer) is a web/...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://mev.tm4.org/#/welcome</td>\n",
       "      <td>https://github.com/dfci-cccb/mev</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>XNAT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>XNAT is an open source imaging informatics pla...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://www.xnat.org</td>\n",
       "      <td>https://bitbucket.org/xnatdev/xnat-web</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Cloud-based Image Biomarker Platform  (C-BIBOP)</td>\n",
       "      <td>C-BIBOP</td>\n",
       "      <td>C-BIBOP is a cloud based platform for medical ...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://cbibop.github.io</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>ivyGlimpse</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bioconductor-based interface to Ivy-GAP (gliob...</td>\n",
       "      <td>Imaging, Omics</td>\n",
       "      <td>https://vjcitn.shinyapps.io/ivyglimpse/</td>\n",
       "      <td>https://github.com/vjcitn/ivygapSE</td>\n",
       "      <td>https://www.youtube.com/watch?v=4qJaodLEI5o</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>The Cancer Proteome Atlas</td>\n",
       "      <td>NaN</td>\n",
       "      <td>The Cancer Proteome Atals is a comprehensive b...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://tcpaportal.org</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>DINC 2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DINC 2.0 is a parallelized meta-docking method...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://dinc.kavrakilab.org</td>\n",
       "      <td>http://dinc.kavrakilab.org/help/</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>AMARETTO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AMARETTO for multiOmics data fusion and reform...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>https://personal.broadinstitute.org/npochet/re...</td>\n",
       "      <td>https://bitbucket.org/gevaertlab/pancanceramar...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>dcmqi: DICOM for Quantitative Imaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dcmqi is a free, open source library that impl...</td>\n",
       "      <td>Data Standards, Imaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://github.com/qiicr/dcmqi</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Tools for Quantitative Analysis of PET Imaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A collection of extensions for 3D Slicer to en...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://qin.iibi.uiowa.edu</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>CLAMP-Cancer</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A tool to quickly build customized natural lan...</td>\n",
       "      <td>Clinical</td>\n",
       "      <td>http://clamp.uth.edu/cancer.php</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Tumor Heterogenity Research Interactive Visual...</td>\n",
       "      <td>THRIVE</td>\n",
       "      <td>A platform for quantitative evaluation of intr...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://www.csb.pitt.edu/ith/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>PETCT-CoSeg</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A PET-CT co-segmentation tool for tumor deline...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://github.com/IOWA-PETCT-COSEG/PETCT_Slic...</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Single Cell Genome Viewer (SCGV)</td>\n",
       "      <td>SCGV</td>\n",
       "      <td>Graphics user interface for single-cell genomi...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://github.com/KrasnitzLab/SCGV</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>BioMuta</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cancer associated gene mutation\\r\\n</td>\n",
       "      <td>Omics</td>\n",
       "      <td>https://hive.biochemistry.gwu.edu/biomuta</td>\n",
       "      <td>https://hive.biochemistry.gwu.edu/biomuta/readme</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>The Cancer Imaging Archive (TCIA)</td>\n",
       "      <td>TCIA</td>\n",
       "      <td>TCIA is NCI’s repository for publicly shared c...</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>http://www.cancerimagingarchive.net</td>\n",
       "      <td>NaN</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>cBioPortal for Cancer Genomics</td>\n",
       "      <td>cBioPortal</td>\n",
       "      <td>The cBioPortal for Cancer Genomics provides vi...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://www.cbioportal.org</td>\n",
       "      <td>https://github.com/cBioPortal/</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Patient-Derived Xenograft (PDX) Finder</td>\n",
       "      <td>PDX Finder</td>\n",
       "      <td>PDX Finder, currently in the prototyping phase...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://www.pdxfinder.org</td>\n",
       "      <td>https://github.com/PDXFinder/pdxfinder</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>The BioMedical Evidence Graph (BMEG)</td>\n",
       "      <td>BMEG</td>\n",
       "      <td>The BMEG is a Cancer Data integration Platform...</td>\n",
       "      <td>Omics</td>\n",
       "      <td>http://bmeg.io</td>\n",
       "      <td>https://github.com/bmeg</td>\n",
       "      <td>https://projectreporter.nih.gov/project_info_d...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Title Tool Acronym  \\\n",
       "0                                           3D Slicer          NaN   \n",
       "1   Allele-Specific Alternative mRNA processing (A...        ASARP   \n",
       "2   Apache Clinical Text and Knowledge Extraction ...       cTAKES   \n",
       "3                                        Bioconductor          NaN   \n",
       "4                Cancer and Phenomics Toolkit (CaPTK)        CaPTK   \n",
       "5                 Cancer Slide Digital Archive (CDSA)         CDSA   \n",
       "6   Cancer-Related Analysis of Variants Toolkit (C...       CRAVAT   \n",
       "7                                            Cistrome          NaN   \n",
       "8   Clinical Interpretation of Variants in Cancer ...        CIViC   \n",
       "9                    Cores Of Recurrent Events (CORE)         CORE   \n",
       "10                                           D2Refine          NaN   \n",
       "11                              DICOM ToolKit (DCMTK)        DCMTK   \n",
       "12  Enhancer Linking by Methylation/Expression Rel...        ELMER   \n",
       "13                                         Feature DB          NaN   \n",
       "14                                Galaxy-P multiOmics     Galaxy-P   \n",
       "15                                        GenePattern          NaN   \n",
       "16  Genome-Independent Identification of RNA Editi...       GIREMI   \n",
       "17                  Integrative Genomics Viewer (IGV)          IGV   \n",
       "18                                           LSTM-CRF          NaN   \n",
       "19                     Methylation Integration (Mint)         Mint   \n",
       "20     Modified UMS, Modified SemRep and SemMedDB-UTH          NaN   \n",
       "21  Mutated Positions in Annotated Structures (MuPIT)        MuPIT   \n",
       "22                   The Network Data Exchange (NDEx)         NDEx   \n",
       "23                Next Generation Clustered Heat Maps          NaN   \n",
       "24                                         OmniSearch          NaN   \n",
       "25                Ontology for MIcroRNA Target (OMIT)         OMIT   \n",
       "26                     LesionTracker Web Image Viewer          NaN   \n",
       "27                                             owl-qa          NaN   \n",
       "28                                             P-Mart          NaN   \n",
       "29                        Personalized Cancer Therapy          NaN   \n",
       "30                                           PhenStat          NaN   \n",
       "31           Quantitative Imaging In Pathology (QUIP)         QUIP   \n",
       "32                                       RADIOMICS.io          NaN   \n",
       "33                                   Region Templates          NaN   \n",
       "34                                          SampleMap          NaN   \n",
       "35      Pathology Imaging Informatics Platform (PIIP)         PIIP   \n",
       "36                                          FHIR ShEx          NaN   \n",
       "37                                         SlicerDMRI          NaN   \n",
       "38          Text Information Extraction System (TIES)         TIES   \n",
       "39  The Laboratory for Individualized Breast Radio...        LIBRA   \n",
       "40                                            Trinity          NaN   \n",
       "41                                          UCSC Xena          NaN   \n",
       "42                                             WebMeV          NaN   \n",
       "43                                               XNAT          NaN   \n",
       "44    Cloud-based Image Biomarker Platform  (C-BIBOP)      C-BIBOP   \n",
       "45                                         ivyGlimpse          NaN   \n",
       "46                          The Cancer Proteome Atlas          NaN   \n",
       "47                                           DINC 2.0          NaN   \n",
       "48                                           AMARETTO          NaN   \n",
       "49              dcmqi: DICOM for Quantitative Imaging          NaN   \n",
       "50     Tools for Quantitative Analysis of PET Imaging          NaN   \n",
       "51                                       CLAMP-Cancer          NaN   \n",
       "52  Tumor Heterogenity Research Interactive Visual...       THRIVE   \n",
       "53                                        PETCT-CoSeg          NaN   \n",
       "54                   Single Cell Genome Viewer (SCGV)         SCGV   \n",
       "55                                            BioMuta          NaN   \n",
       "56                 The Cancer Imaging Archive (TCIA)          TCIA   \n",
       "57                    cBioPortal for Cancer Genomics    cBioPortal   \n",
       "58            Patient-Derived Xenograft (PDX) Finder    PDX Finder   \n",
       "59              The BioMedical Evidence Graph (BMEG)          BMEG   \n",
       "\n",
       "                                                 Body  \\\n",
       "0   3D Slicer is the free open source software for...   \n",
       "1   A software pipeline for prediction of allele-s...   \n",
       "2   The tool extracts deep phenotypic information ...   \n",
       "3   Bioconductor provides tools for the analysis a...   \n",
       "4   CaPTk is a tool that facilitates translation o...   \n",
       "5   The CDSA is a web-based platform to support th...   \n",
       "6   CRAVAT is an easy to use web-based tool for an...   \n",
       "7   Curated and processed human/mouse ChIP/DNase-s...   \n",
       "8   CIViC is an open access, open source, communit...   \n",
       "9   CORE is a statistically supported computationa...   \n",
       "10  D2Refine is an integrated platform that enable...   \n",
       "11         DCMTK is an open source DICOM toolkit.\\r\\n   \n",
       "12  R tool for analysis of DNA methylation and exp...   \n",
       "13  The backend database to manage segmentation re...   \n",
       "14  A unified platform for integrative genomic-pro...   \n",
       "15  GenePattern is a genomic analysis platform tha...   \n",
       "16  Identification of single-nucleotide variants i...   \n",
       "17  The Integrative Genomics Viewer (IGV) is a hig...   \n",
       "18  LSTM-CRF uses Natural Language Processing meth...   \n",
       "19  A comprehensive software pipeline and set of G...   \n",
       "20  Modified UMLS, modified SemRep and SemMedDB-UT...   \n",
       "21  MuPIT provides 3D visualization of mutation cl...   \n",
       "22  NDEx is an online commons where scientists can...   \n",
       "23  Next-Generation (Clustered) Heat Maps are inte...   \n",
       "24  OmniSearch is a semantic search software based...   \n",
       "25  OMIT is a domain ontology specifically designe...   \n",
       "26  Extensible open-source zero-footprint web imag...   \n",
       "27  owl-qa is an OWL-based QA tool for cancer stud...   \n",
       "28  Web-based software platform to enable analysis...   \n",
       "29  Curated website cataloguing clinically actiona...   \n",
       "30  PhenStat is a freely available R package that ...   \n",
       "31  This site hosts web accessible applications, t...   \n",
       "32  RADIOMICS.io is a open source platform for inf...   \n",
       "33  A suite of libraries and runtime to support im...   \n",
       "34  Projects samples into a GoogleMap explorer to ...   \n",
       "35  Freely downloadable pathology viewer which sup...   \n",
       "36  ShEx is a QA tool based on the Shape Expressio...   \n",
       "37  Diffusion magnetic resonance imaging for neuro...   \n",
       "38  TIES is a service based software system for ac...   \n",
       "39  LIBRA is a fully-automatic breast density esti...   \n",
       "40  Trinity Cancer Transcriptome Analysis Toolkit ...   \n",
       "41  UCSC Xena securely analyzes and visualizes you...   \n",
       "42  Web MeV (Multiple-experiment Viewer) is a web/...   \n",
       "43  XNAT is an open source imaging informatics pla...   \n",
       "44  C-BIBOP is a cloud based platform for medical ...   \n",
       "45  Bioconductor-based interface to Ivy-GAP (gliob...   \n",
       "46  The Cancer Proteome Atals is a comprehensive b...   \n",
       "47  DINC 2.0 is a parallelized meta-docking method...   \n",
       "48  AMARETTO for multiOmics data fusion and reform...   \n",
       "49  dcmqi is a free, open source library that impl...   \n",
       "50  A collection of extensions for 3D Slicer to en...   \n",
       "51  A tool to quickly build customized natural lan...   \n",
       "52  A platform for quantitative evaluation of intr...   \n",
       "53  A PET-CT co-segmentation tool for tumor deline...   \n",
       "54  Graphics user interface for single-cell genomi...   \n",
       "55                Cancer associated gene mutation\\r\\n   \n",
       "56  TCIA is NCI’s repository for publicly shared c...   \n",
       "57  The cBioPortal for Cancer Genomics provides vi...   \n",
       "58  PDX Finder, currently in the prototyping phase...   \n",
       "59  The BMEG is a Cancer Data integration Platform...   \n",
       "\n",
       "                   Category  \\\n",
       "0                   Imaging   \n",
       "1                     omics   \n",
       "2                  Clinical   \n",
       "3                     Omics   \n",
       "4                   Imaging   \n",
       "5                   Imaging   \n",
       "6                     Omics   \n",
       "7                     Omics   \n",
       "8           Omics, Clinical   \n",
       "9                     Omics   \n",
       "10           Data Standards   \n",
       "11  Imaging, Data Standards   \n",
       "12                    Omics   \n",
       "13                  Imaging   \n",
       "14                    Omics   \n",
       "15                    Omics   \n",
       "16                    Omics   \n",
       "17                    Omics   \n",
       "18                    Omics   \n",
       "19                    Omics   \n",
       "20          Clinical, Omics   \n",
       "21                    Omics   \n",
       "22          Network Biology   \n",
       "23                    Omics   \n",
       "24           Data Standards   \n",
       "25           Data Standards   \n",
       "26                  Imaging   \n",
       "27           Data Standards   \n",
       "28                    Omics   \n",
       "29          Clinical, Omics   \n",
       "30                    Omics   \n",
       "31                  Imaging   \n",
       "32                  Imaging   \n",
       "33                  Imaging   \n",
       "34                    Omics   \n",
       "35                  Imaging   \n",
       "36           Data Standards   \n",
       "37                  Imaging   \n",
       "38                 Clinical   \n",
       "39                  Imaging   \n",
       "40                    Omics   \n",
       "41                    Omics   \n",
       "42                    Omics   \n",
       "43                  Imaging   \n",
       "44                  Imaging   \n",
       "45           Imaging, Omics   \n",
       "46                    Omics   \n",
       "47                    Omics   \n",
       "48                    Omics   \n",
       "49  Data Standards, Imaging   \n",
       "50                  Imaging   \n",
       "51                 Clinical   \n",
       "52                  Imaging   \n",
       "53                  Imaging   \n",
       "54                    Omics   \n",
       "55                    Omics   \n",
       "56                  Imaging   \n",
       "57                    Omics   \n",
       "58                    Omics   \n",
       "59                    Omics   \n",
       "\n",
       "                                             Tool URL  \\\n",
       "0                              https://www.slicer.org   \n",
       "1   https://www.ibp.ucla.edu/research/xiao/Softwar...   \n",
       "2                            http://ctakes.apache.org   \n",
       "3                        https://www.bioconductor.org   \n",
       "4                     http://captk.projects.nitrc.org   \n",
       "5               http://cancer.digitalslidearchive.net   \n",
       "6                        http://www.cravat.us/CRAVAT/   \n",
       "7   http://cistrome.org/Cistrome/Cistrome_Project....   \n",
       "8                 https://civic.genome.wustl.edu/home   \n",
       "9                                                 NaN   \n",
       "10          https://github.com/caCDE-QA/D2Refine/wiki   \n",
       "11                 http://dicom.offis.de/dcmtk.php.en   \n",
       "12  http://bioconductor.org/packages/release/bioc/...   \n",
       "13                                                NaN   \n",
       "14                                 http://galaxyp.org   \n",
       "15  http://software.broadinstitute.org/cancer/soft...   \n",
       "16                                                NaN   \n",
       "17                                 http://www.igv.org   \n",
       "18                                                NaN   \n",
       "19                  https://github.com/sartorlab/mint   \n",
       "20   https://skr3.nlm.nih.gov/SemMedDB/index_uth.html   \n",
       "21        http://mupit.icm.jhu.edu/MuPIT_Interactive/   \n",
       "22                            http://www.ndexbio.org/   \n",
       "23  http://bioinformatics.mdanderson.org/TCGA/NGCH...   \n",
       "24         http://omnisearch.soc.southalabama.edu/ui/   \n",
       "25  http://omnisearch.soc.southalabama.edu/w/index...   \n",
       "26                      http://lesiontracker.ohif.org   \n",
       "27                 https://github.com/caCDE-QA/owl-qa   \n",
       "28              https://pmart.labworks.org/index.html   \n",
       "29                      https://pct.mdanderson.org/#/   \n",
       "30                                                NaN   \n",
       "31                     http://quip.bmi.stonybrook.edu   \n",
       "32                           https://www.radiomics.io   \n",
       "33                                                NaN   \n",
       "34                          https://tumormap.ucsc.edu   \n",
       "35                                 http://pathiip.org   \n",
       "36                   https://github.com/caCDE-QA/ShEX   \n",
       "37                             http://dmri.slicer.org   \n",
       "38                          http://ties.dbmi.pitt.edu   \n",
       "39          https://www.med.upenn.edu/sbia/libra.html   \n",
       "40  https://github.com/trinityrnaseq/trinityrnaseq...   \n",
       "41                               http://xena.ucsc.edu   \n",
       "42                       http://mev.tm4.org/#/welcome   \n",
       "43                                http://www.xnat.org   \n",
       "44                            http://cbibop.github.io   \n",
       "45            https://vjcitn.shinyapps.io/ivyglimpse/   \n",
       "46                              http://tcpaportal.org   \n",
       "47                         http://dinc.kavrakilab.org   \n",
       "48  https://personal.broadinstitute.org/npochet/re...   \n",
       "49                                                NaN   \n",
       "50                          http://qin.iibi.uiowa.edu   \n",
       "51                    http://clamp.uth.edu/cancer.php   \n",
       "52                       http://www.csb.pitt.edu/ith/   \n",
       "53                                                NaN   \n",
       "54                                                NaN   \n",
       "55          https://hive.biochemistry.gwu.edu/biomuta   \n",
       "56                http://www.cancerimagingarchive.net   \n",
       "57                          http://www.cbioportal.org   \n",
       "58                           http://www.pdxfinder.org   \n",
       "59                                     http://bmeg.io   \n",
       "\n",
       "                                            Code Repo  \\\n",
       "0                           https://github.com/Slicer   \n",
       "1                  https://github.com/cyruschan/ASARP   \n",
       "2              http://ctakes.apache.org/downloads.cgi   \n",
       "3      http://bioconductor.org/developers/how-to/git/   \n",
       "4                                                 NaN   \n",
       "5                                                 NaN   \n",
       "6                https://github.com/KarchinLab/CRAVAT   \n",
       "7                                                 NaN   \n",
       "8                      https://github.com/griffithlab   \n",
       "9     http://cran.us.r-project.org/web/packages/CORE/   \n",
       "10               https://github.com/caCDE-QA/D2Refine   \n",
       "11                               http://git.dcmtk.org   \n",
       "12                  https://github.com/lijingya/ELMER   \n",
       "13     https://github.com/SBU-BMI/pathomics_featuredb   \n",
       "14  https://github.com/galaxyproteomics/tools-galaxyp   \n",
       "15  https://github.com/genepattern/genepattern-server   \n",
       "16                 https://github.com/zhqingit/giremi   \n",
       "17                         https://github.com/igvteam   \n",
       "18      https://github.com/abhyudaynj/LSTM-CRF-models   \n",
       "19                  https://github.com/sartorlab/mint   \n",
       "20                                                NaN   \n",
       "21                                                NaN   \n",
       "22                         https://github.com/ndexbio   \n",
       "23                                                NaN   \n",
       "24                 https://github.com/OmniSearch/ncro   \n",
       "25                 https://github.com/OmniSearch/ncro   \n",
       "26                            https://github.com/OHIF   \n",
       "27                 https://github.com/caCDE-QA/owl-qa   \n",
       "28           https://hub.docker.com/r/pnnl/pmart-web/   \n",
       "29                                                NaN   \n",
       "30  https://bioconductor.org/packages/release/bioc...   \n",
       "31                  https://github.com/SBU-BMI/nscale   \n",
       "32                       https://github.com/Radiomics   \n",
       "33        https://github.com/SBU-BMI/region-templates   \n",
       "34                                                NaN   \n",
       "35                                                NaN   \n",
       "36                   https://github.com/caCDE-QA/ShEX   \n",
       "37                      https://github.com/SlicerDMRI   \n",
       "38           https://sourceforge.net/projects/caties/   \n",
       "39                                                NaN   \n",
       "40     https://github.com/trinityrnaseq/trinityrnaseq   \n",
       "41                       https://github.com/ucscXena/   \n",
       "42                   https://github.com/dfci-cccb/mev   \n",
       "43             https://bitbucket.org/xnatdev/xnat-web   \n",
       "44                                                NaN   \n",
       "45                 https://github.com/vjcitn/ivygapSE   \n",
       "46                                                NaN   \n",
       "47                   http://dinc.kavrakilab.org/help/   \n",
       "48  https://bitbucket.org/gevaertlab/pancanceramar...   \n",
       "49                     https://github.com/qiicr/dcmqi   \n",
       "50                                                NaN   \n",
       "51                                                NaN   \n",
       "52                                                NaN   \n",
       "53  https://github.com/IOWA-PETCT-COSEG/PETCT_Slic...   \n",
       "54                https://github.com/KrasnitzLab/SCGV   \n",
       "55   https://hive.biochemistry.gwu.edu/biomuta/readme   \n",
       "56                                                NaN   \n",
       "57                     https://github.com/cBioPortal/   \n",
       "58             https://github.com/PDXFinder/pdxfinder   \n",
       "59                            https://github.com/bmeg   \n",
       "\n",
       "                                         Project Info  \n",
       "0   https://projectreporter.nih.gov/project_info_d...  \n",
       "1   https://projectreporter.nih.gov/project_info_d...  \n",
       "2   https://projectreporter.nih.gov/project_info_d...  \n",
       "3   https://projectreporter.nih.gov/project_info_d...  \n",
       "4   https://projectreporter.nih.gov/project_info_d...  \n",
       "5   https://projectreporter.nih.gov/project_info_d...  \n",
       "6   https://projectreporter.nih.gov/project_info_d...  \n",
       "7   https://projectreporter.nih.gov/project_info_d...  \n",
       "8   https://projectreporter.nih.gov/project_info_d...  \n",
       "9   https://projectreporter.nih.gov/project_info_d...  \n",
       "10  https://projectreporter.nih.gov/project_info_d...  \n",
       "11  https://projectreporter.nih.gov/project_info_d...  \n",
       "12  https://projectreporter.nih.gov/project_info_d...  \n",
       "13  https://projectreporter.nih.gov/project_info_d...  \n",
       "14  https://projectreporter.nih.gov/project_info_d...  \n",
       "15  https://projectreporter.nih.gov/project_info_d...  \n",
       "16  https://projectreporter.nih.gov/project_info_d...  \n",
       "17  https://projectreporter.nih.gov/project_info_d...  \n",
       "18  https://projectreporter.nih.gov/project_info_d...  \n",
       "19  https://projectreporter.nih.gov/project_info_d...  \n",
       "20  https://projectreporter.nih.gov/project_info_d...  \n",
       "21  https://projectreporter.nih.gov/project_info_d...  \n",
       "22  https://projectreporter.nih.gov/project_info_d...  \n",
       "23  https://projectreporter.nih.gov/project_info_d...  \n",
       "24  https://projectreporter.nih.gov/project_info_d...  \n",
       "25  https://projectreporter.nih.gov/project_info_d...  \n",
       "26  https://projectreporter.nih.gov/project_info_d...  \n",
       "27  https://projectreporter.nih.gov/project_info_d...  \n",
       "28  https://projectreporter.nih.gov/project_info_d...  \n",
       "29  https://projectreporter.nih.gov/project_info_d...  \n",
       "30  https://projectreporter.nih.gov/project_info_d...  \n",
       "31  https://projectreporter.nih.gov/project_info_d...  \n",
       "32  https://projectreporter.nih.gov/project_info_d...  \n",
       "33  https://projectreporter.nih.gov/project_info_d...  \n",
       "34  https://projectreporter.nih.gov/project_info_d...  \n",
       "35  https://projectreporter.nih.gov/project_info_d...  \n",
       "36  https://projectreporter.nih.gov/project_info_d...  \n",
       "37  https://projectreporter.nih.gov/project_info_d...  \n",
       "38  https://projectreporter.nih.gov/project_info_d...  \n",
       "39  https://projectreporter.nih.gov/project_info_d...  \n",
       "40  https://projectreporter.nih.gov/project_info_d...  \n",
       "41  https://projectreporter.nih.gov/project_info_d...  \n",
       "42  https://projectreporter.nih.gov/project_info_d...  \n",
       "43  https://projectreporter.nih.gov/project_info_d...  \n",
       "44  https://projectreporter.nih.gov/project_info_d...  \n",
       "45        https://www.youtube.com/watch?v=4qJaodLEI5o  \n",
       "46  https://projectreporter.nih.gov/project_info_d...  \n",
       "47                                                NaN  \n",
       "48                                                NaN  \n",
       "49  https://projectreporter.nih.gov/project_info_d...  \n",
       "50  https://projectreporter.nih.gov/project_info_d...  \n",
       "51                                                NaN  \n",
       "52  https://projectreporter.nih.gov/project_info_d...  \n",
       "53  https://projectreporter.nih.gov/project_info_d...  \n",
       "54  https://projectreporter.nih.gov/project_info_d...  \n",
       "55                                                NaN  \n",
       "56  https://projectreporter.nih.gov/project_info_d...  \n",
       "57  https://projectreporter.nih.gov/project_info_d...  \n",
       "58                                                NaN  \n",
       "59  https://projectreporter.nih.gov/project_info_d...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_path = '../resources/itcr-conn-map-v2.0_nodes.txt'\n",
    "node_data = pd.read_csv(tsv_path, sep=\"\\t\")\n",
    "node_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Annotate the nodes in the network according to the plan.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_to_node_map = {}\n",
    "for index, node in itcr_network.get_nodes():\n",
    "    name_to_node_map[node.get_name()] = node\n",
    "        \n",
    "for index, row in node_data.iterrows():\n",
    "    # Identify the corresponding node (skip if there is no such node)\n",
    "    node = name_to_node_map.get(row['Title']) or name_to_node_map.get(row['Tool Acronym'])\n",
    "    if node:\n",
    "        \n",
    "        # Add attributes to the node\n",
    "        attributes = [\"Category\", \"Tool URL\", \"Code Repo\", \"Project Info\"]\n",
    "        for att in attributes:\n",
    "            itcr_network.set_node_attribute(node, att, row[att])\n",
    "            \n",
    "        # Set the \"Tool Description\" attribute of the node to the value of the \"Body\" column\n",
    "        itcr_network.set_node_attribute(node, \"Tool Description\", row['Body'])\n",
    "        \n",
    "        # Set the node name to be the value of the \"Title\" column\n",
    "        #node.set_name(row[\"Title\"])\n",
    "        \n",
    "        # Set the \"label\" attribute to the value of the \"Tool Acronym\" when available, \n",
    "        # otherwise use the value of the \"Title\" column.\n",
    "        acronym = row[\"Tool Acronym\"]\n",
    "        if not isinstance(acronym, str):\n",
    "            acronym = None\n",
    "        itcr_network.set_node_attribute(node, \"label\", acronym or row[\"Title\"] or \"no label\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Spot-check the network by printing out a sample of the nodes in the modified network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name: ITCR Connectivity Map\n",
      "Nodes: 36\n",
      "Edges: 81\n",
      "Node Attributes: 174\n",
      "Edge Attributes: 241\n",
      "\n",
      "TIES: TIES = TIES is a service based software system for acquiring, deidentifying, and processing clinical text reports using natural language processing, and also for querying, sharing and using this data to foster tissue and image based research, within and between institutions.\r\n",
      "\n",
      "TCPA: None = None\n",
      "HIVE (BioExpress): None = None\n",
      "TCIA: TCIA = TCIA is NCI’s repository for publicly shared cancer imaging data. TCIA collections include radiology and pathology images, clinical and clinical trial data, image derived annotations and quantitative features and a growing collection of related ‘omics data both from clinical and pre-clinical studies.\r\n",
      "\n",
      "OmniSearch: OmniSearch = OmniSearch is a semantic search software based upon the OMIT ontology. While OmniSearch is by its nature extensible, its initial focus is in human cancer research.\r\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(itcr_network.get_summary())\n",
    "\n",
    "for id, node in itertools.islice(itcr_network.get_nodes(), 5):\n",
    "    label = itcr_network.get_node_attribute(node, \"label\")\n",
    "    tool = itcr_network.get_node_attribute(node, \"Tool Description\")\n",
    "    print(\"%s: %s = %s\" % (node.get_name(), label, tool))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Upload the network to your NDEx account as a new network.\n",
    "\n",
    "If you started by loading an one of your networks (such as a clone of the example network), then you can *update* it instead, overwriting its content. \n",
    "\n",
    "In the commented line of below, the optional parameter *update_uuid* specifies the update behavior."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "upload_message = my_network.upload(my_server, my_account, my_password)\n",
    "# upload_message = my_network.upload(my_server, my_account, my_password, update_uuid=my_network_uuid)\n",
    "print(upload_message)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
